drivers: serial: lis3mdl: support for SPI

DeMurlot opened the pull request:
This pull request add two things to the lis3mdl driver
- support for usage through SPI
- Kconfig option to enable die temperature reading

The first part is heavily inspired by the lsm6dsl's way of supporting both I2C and SPI.

DeMurlot commented:
The current Kconfig default is to disable support for die temperature measurement and breaks backwards comparability. Should I enable it by default?

jeppenodgaard commented:
> The current Kconfig default is to disable support for die temperature measurement and breaks backwards comparability. Should I enable it by default?

I think it is safest to keep it as is.

If it is changed I think that change should have its own commit, and maybe probably be mentioned in the 4.0 migration guide.

jeppenodgaard commented:
The commits needs a body as described in:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines

avisconti commented on drivers/sensor/st/lis3mdl/lis3mdl.c:
```diff
@@ -85,11 +91,12 @@ int lis3mdl_sample_fetch(const struct device *dev, enum sensor_channel chan)
 		LOG_DBG("Failed to fetch temperature sample.");
 		return -EIO;
 	}
+	drv_data->temp_sample = sys_le16_to_cpu(buf[3]);
```
where did you read these two bytes?

avisconti reviewed (changes requested):
The first commit looks ok a part a couple of comments

avisconti commented on dts/bindings/sensor/st,lis3mdl-spi.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2017, Linaro Limited
```
2024 and maybe not Linaro

DeMurlot replied:
fixed, thank you

avisconti commented on dts/bindings/sensor/st,lis3mdl-i2c.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2017, Linaro Limited
```
2024 and Linaro?

DeMurlot replied:
fixed

avisconti commented on drivers/sensor/st/lis3mdl/lis3mdl.c:
```diff
@@ -172,17 +174,73 @@ int lis3mdl_init(const struct device *dev)
+	static struct lis3mdl_data lis3mdl_data_##inst;			\
+	static struct lis3mdl_config lis3mdl_config_##inst =	\
+		LIS3MDL_CONFIG_I2C(inst);				\
+	LIS3MDL_DEVICE_INIT(inst)
```
this macro is common between i2c and spi, so maybe you can call only once in ```LIS3MDL_DEFINE()``` as some other st driver do.

avisconti commented on drivers/sensor/st/lis3mdl/lis3mdl.c:
```diff
@@ -172,17 +174,73 @@ int lis3mdl_init(const struct device *dev)
+		.bus_init = lis3mdl_i2c_init,				\
+		.bus_cfg.i2c = I2C_DT_SPEC_INST_GET(inst),		\
+		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, irq_gpios),	\
+		(LIS3MDL_CFG_IRQ(inst)), ())				\
```
same for i2c and spi. Maybe call only once?

avisconti reviewed (changes requested):
maybe we could move to stmemsc directly instead of introducing hw_tf?

avisconti commented:
@DeMurlot I guess you may also want to add the spi test for the sensor in ```tests/drivers/build_all/sensor/```

Commit:
drivers: sensor: LIS3MDL: make die temperature reading optional

The LIS3MDL sensor driver is slowed down by an additional
transaction to fetch the die temperature. This commit allowes to
enable/disable this via a KConfig option. Default is enables in
order to preserve backwards compatability.

Signed-off-by: De Murlot <demurlotpierre@gmail.com>

Commit:
drivers: sensor: LIS3MDL: support SPI

The LIS3MDL sensor driver does not support SPI.
This commit adds support for usage through SPI.

Signed-off-by: De Murlot <demurlotpierre@gmail.com>
